File: 2025/06 June/02 Unique Paths in a Grid/main.py
# Link: https://www.geeksforgeeks.org/problems/unique-paths-in-a-grid--170647/1


class Solution:
    def uniquePaths(self, grid):
        r, c = len(grid), len(grid[0])
        dp = [0] * c

        dp[0] = 1 if grid[0][0] == 0 else 0

        for i in range(r):
            for j in range(c):
                if grid[i][j] == 1:
                    dp[j] = 0
                elif j > 0:
                    dp[j] += dp[j - 1]

        return dp[c - 1]


if __name__ == "__main__":
    obj = Solution()
    grid = [[0, 0, 0], [0, 1, 0], [0, 0, 0]]
    print(obj.uniquePaths(grid))

    grid = [[1, 0, 1]]
    print(obj.uniquePaths(grid))

File: 2025/06 June/02 Unique Paths in a Grid/main.cpp
// Link: https://www.geeksforgeeks.org/problems/unique-paths-in-a-grid--170647/1

#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    int uniquePaths(vector<vector<int>> &grid)
    {
        int r = grid.size(), c = grid[0].size();
        vector<int> dp(c, 0);

        dp[0] = (grid[0][0] == 0) ? 1 : 0;
        for (int i = 0; i < r; i++)
        {
            for (int j = 0; j < c; j++)
            {
                if (grid[i][j] == 1)
                {
                    dp[j] = 0;
                }
                else if (j > 0)
                {
                    dp[j] = (dp[j] + dp[j - 1]);
                }
            }
        }

        return dp[c - 1];
    }
};

int main()
{
    Solution sol;
    vector<vector<int>> grid = {
        {0, 0, 0},
        {0, 1, 0},
        {0, 0, 0}};
    cout << "Unique Paths: " << sol.uniquePaths(grid) << endl;

    grid = {{1, 0, 1}};
    cout << "Unique Paths: " << sol.uniquePaths(grid) << endl;
    return 0;
}